types/bluebird/index.d.ts

```diff
@@ -44,10 +45,8 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
   /**
    * Promises/A+ `.then()`. Returns a new promise chained from this promise. The new promise will be rejected or resolved dedefer on the passed `fulfilledHandler`, `rejectedHandler` and the state of this promise.
    */
-  then<U1, U2>(onFulfill: (value: R) => U1 | Bluebird.Thenable<U1>, onReject: (error: any) => U2 | Bluebird.Thenable<U2>): Bluebird<U1 | U2>;
-  then<U>(onFulfill: (value: R) => U | Bluebird.Thenable<U>, onReject: (error: any) => U | Bluebird.Thenable<U>): Bluebird<U>;
-  then<U>(onFulfill: (value: R) => U | Bluebird.Thenable<U>): Bluebird<U>;
-  then(): Bluebird<R>;
+  // Based on PromiseLike.then, but returns a Bluebird instance.
+  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): Bluebird<TResult1 | TResult2>;
```

Comment:
@lhecker Oops. That's because `{ a: 1 }` and `{ a: 2 }` create anonymous types that are not `===`. They would be eliminated by subtype reduction, but we rarely do subtype reduction anymore to get rid of cases like `result1`. Still, the types of `result1` and `result2` should be interchangeable. The only differences (I think) are:
 
1. In printing the type, which should be fixed by making `typeToString` do subtype reduction.
2. In possible compile-time performance, if the compiler ends up creating giant unions by mistake.

Both these could be avoided by doing subtype reduction every time the compiler *creates* a union, but that has upfront costs and hurts correctness sometimes as well.

Point taken about IDE usability. If real code would show an ugly union right now, the overloads are superior. I don't know whether the two callbacks are returning anonymous types as opposed to type references that would be `===` inside the compiler.